source/common/ssl/context_config_impl.cc

```diff
@@ -16,21 +16,36 @@ namespace Ssl {
 
 namespace {
 
-template <typename T>
-const std::shared_ptr<T> findSecretAndCheckType(const Secret::SecretManager& secret_manager,
-                                                const std::string& name) {
-
-  auto secret = secret_manager.findSecret(name);
-  if (secret == nullptr) {
-    throw EnvoyException(fmt::format("Static secret is not defined: {}", name));
+const std::string readCrtificateChain(const envoy::api::v2::auth::CommonTlsContext& config,
+                                      Secret::SecretManager& secret_manager) {
+  if (!config.tls_certificates().empty()) {
+    return Config::DataSource::read(config.tls_certificates()[0].certificate_chain(), true);
+  } else if (!config.tls_certificate_sds_secret_configs().empty()) {
+    auto name = config.tls_certificate_sds_secret_configs()[0].name();
+    auto secret = secret_manager.findSecret<Ssl::TlsCertificateConfigImpl>(name);
+    if (!secret) {
+      throw EnvoyException(fmt::format("Static secret is not defined: {}", name));
+    }
+    return secret->certificateChain();
+  } else {
+    return std::string("");
   }
+}
 
-  auto resp = std::dynamic_pointer_cast<T>(secret);
-  if (resp == nullptr) {
-    throw EnvoyException(fmt::format("Invalid type of secret was returned for '{}'", name));
+const std::string readPrivateKey(const envoy::api::v2::auth::CommonTlsContext& config,
+                                 Secret::SecretManager& secret_manager) {
+  if (!config.tls_certificates().empty()) {
+    return Config::DataSource::read(config.tls_certificates()[0].private_key(), true);
+  } else if (!config.tls_certificate_sds_secret_configs().empty()) {
+    auto name = config.tls_certificate_sds_secret_configs()[0].name();
+    auto secret = secret_manager.findSecret<Ssl::TlsCertificateConfigImpl>(name);
+    if (!secret) {
+      throw EnvoyException(fmt::format("Static secret is not defined: {}", name));
+    }
+    return secret->privateKey();
+  } else {
+    return std::string("");
```

Comment:
These two functions are similar.  like to combine them.  How about this:

string readConfig(config, mgr,  READ_TCL_CERT_FN,  READ_CONFIG_IMPL)  {
if (!config.tls_certificates().empty()) {
   return Config::DataSource::read(READ_TCL_CERT_FN(config.tls_certificates()[0])), true);  } else if (!config.tls_certificate_sds_secret_configs().empty()) {
+    
    return READ_CONFIG_IMPL(second_impl);
}

typedef std::function<const DataStore& (const tcl_certificate&)>  read_tcl_cert_fn;